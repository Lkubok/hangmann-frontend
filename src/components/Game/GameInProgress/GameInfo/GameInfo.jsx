import React, { Component } from "react";
import { connect } from "react-redux";
import * as selectors from "../../../../reducers/selectors";
import "./GameInfo.scss";
import { closeGame, saveTime } from "../../../../actions/gameActions";
const { REACT_APP_GAME_TIME } = process.env;

export class GameInfo extends Component {
  constructor(props) {
    super(props);
    this.state = {
      startTime: this.props.startTime,
      timeLeft: this.props.stoppedTime
        ? this.props.stoppedTime
        : parseInt(this.props.startTime) + parseInt(REACT_APP_GAME_TIME)
    };
  }
  componentDidMount() {
    this.interval = setInterval(() => {
      this.setState(prevState => {
        return { timeLeft: prevState.timeLeft - 1000 };
      });
    }, 1000);
  }
  componentDidUpdate() {
    const { startTime, timeLeft } = this.state;
    const { isFinished, lifes } = this.props;
    if (timeLeft < startTime || lifes === 0 || isFinished) {
      clearInterval(this.interval);
      this.props.closeGame();
    }
  }
  componentWillUnmount() {
    const { saveTime } = this.props;
    saveTime(this.state.timeLeft);
    clearInterval(this.interval);
  }
  showTime = () => {
    const { startTime, timeLeft } = this.state;
    const timeLeftValue = timeLeft - startTime;
    return timeLeftValue > 0 ? timeLeftValue / 1000 + "s" : "0s";
  };
  getStartDate = date => {
    const startTime = new Date(date);
    return `${startTime.getHours()} : ${startTime.getMinutes()} : ${startTime.getSeconds()}`;
  };
  render() {
    const { startTime, lifes, stateOfGame } = this.props;
    return (
      <div className="game-info">
        <div className="info-elem">
          <p className>
            Lifes left: <span>{lifes}</span>
          </p>
        </div>
        <div className="info-elem">
          <p className>
            Game Start Time: <span>{this.getStartDate(startTime)}</span>
          </p>
        </div>
        <div className="info-elem">
          <p className>
            State Of Game: <span>{stateOfGame}</span>
          </p>
        </div>
        <div className="info-elem">
          <p className>
            Time Left: <span>{this.showTime()}</span>
          </p>
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  lifes: selectors.getLifes(state),
  startTime: selectors.getStartTime(state),
  stateOfGame: selectors.getStateOfGame(state),
  isFinished: selectors.getIsFinished(state),
  stoppedTime: selectors.getTimeLeft(state)
});
const mapDispatchToProps = {
  closeGame,
  saveTime
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(GameInfo);
