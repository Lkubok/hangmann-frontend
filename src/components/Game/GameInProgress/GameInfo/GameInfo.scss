@import "../../../../scss/mixins.scss";
@import "../../../../scss/variables.scss";

.game-info {
  @include flex-cc-row;
  flex-basis: 100%;
  justify-content: space-around;
  margin-top: 40px;
  margin-bottom: 0px;
  .info-elem {
    @include flex-cc-row;
    flex-basis: 33%;
    padding: 20px;
    p {
      span {
        color: $primary-color;
      }
    }
  }
}
