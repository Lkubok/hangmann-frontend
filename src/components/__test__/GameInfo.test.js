import React from "react";
import Enzyme, { shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import { GameInfo } from "../Game/GameInProgress/GameInfo/GameInfo";
Enzyme.configure({ adapter: new Adapter() });

describe("GameInfo component", () => {
  beforeEach(() => {});
  it("Should match snapshot", () => {
    const wrapper = shallow(<GameInfo />);
    expect(wrapper).toMatchSnapshot();
  });

  it("Should launch interval when mounted", () => {
    const wrapper = shallow(<GameInfo />);
    const intervalSpy = jest.fn();
    wrapper.instance().launchInterval = intervalSpy;
    wrapper.instance().forceUpdate();
  });
});
