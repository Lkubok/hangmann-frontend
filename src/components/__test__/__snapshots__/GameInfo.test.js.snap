// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameInfo component Should match snapshot 1`] = `
<div
  className="game-info"
>
  <div
    className="info-elem"
  >
    <p>
      Lifes left: 
      <span />
    </p>
  </div>
  <div
    className="info-elem"
  >
    <p>
      Game Start Time: 
      <span>
        NaN : NaN : NaN
      </span>
    </p>
  </div>
  <div
    className="info-elem"
  >
    <p>
      State Of Game: 
      <span />
    </p>
  </div>
  <div
    className="info-elem"
  >
    <p>
      Time Left: 
      <span>
        0s
      </span>
    </p>
  </div>
</div>
`;
